### PR TITLE
feat: amplify console hosting plugin

### PR DESCRIPTION
**Pull request by garyleefight**
hosting plugin with e2e test cases

### CODE REVIEW

File: packages/amplify-console-hosting/amplify-plugin.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    "commands": [
+        "add",
+        "configure",
+        "remove",
+        "publish",
+        "help",
+        "status",
+        "serve"
+    ],
+    "displayName": "Hosting with Amplify Console (Managed Hosting, custom domains, continuous deployment)",
```
Comment (swaminator) [inline packages/amplify-console-hosting/amplify-plugin.json]: Hosting with Amplify Console (Managed hosting with custom domains, Continuous deployment)

File: packages/amplify-console-hosting/commands/help.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+async function run(context) {
+  // print out the help message of your plugin
+  context.print.info('help command to be implemented.');
```
Comment (swaminator) [inline packages/amplify-console-hosting/commands/help.js]: amplify hosting serve: Opens your deployed site

amplify hosting configure: Set up custom domains, redirects, password protection, and more via the Amplify Console.

amplify publish: Publishes changes to manually deployed apps. For continuous deployment, please push to your git branch to deploy updates.

amplify hosting remove: Remove hosting from you app.

File: packages/amplify-console-hosting/constants/question-constants.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import chalk from 'chalk';
+
+export const DEPLOY_TYPE_QUESTION = `Choose a ${chalk.red('type')}`;
+export const DEPLOY_TYPE_QUESTION_MANUAL = 'Manual deployment';
+export const DEPLOY_TYPE_QUESTION_CICD = 'Continuous deployment (Git-based deployments)';
+export const LEARN_MORE = 'Learn more';
+
+export const CICD_CONFIRM_QUESTION = 'Continuous deployment is configured in the Amplify Console. Please hit enter once you connect your repository';
+
+export const VIEW_APP_QUESTION = `You have set up CI/CD with Amplify Console. \
```
Comment (swaminator) [inline packages/amplify-console-hosting/constants/question-constants.js]: You have set up continuous deployment with Amplify Console.

File: packages/amplify-console-hosting/constants/question-constants.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+Run ${chalk.green('git push')} from a connected branch to publish updates. \
+Open your Amplify Console app to view connected branches?`;
+
+export const APP_CONFIGURE_QUESTION = `We recommends you open ${chalk.green('AWS Amplify Console')} to configure advanced settings such as ${chalk.yellow('rewrite/redirect, basic auth, custom domain...')} Open your Amplify Console app to configure?`;
```
Comment (swaminator) [inline packages/amplify-console-hosting/constants/question-constants.js]: Configure settings such as custom domains, redirects, and password protection using the Amplify Console. Continue?

File: packages/amplify-console-hosting/constants/question-constants.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+export const APP_CICD_SERVE_QUESTION = 'You have set up Continuous deployment with Amplify Console. Visit Amplify Console to view your frontend URL?';
```
Comment (swaminator) [inline packages/amplify-console-hosting/constants/question-constants.js]: You have set up continuous deployment with Amplify Console. Open your Amplify Console app to view connected branches?

File: packages/amplify-console-hosting/constants/question-constants.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+export const APP_MANUAL_SERVE_QUESTION = 'You have set up Manual deployment with Amplify Console. Visit Amplify Console to view your frontend URL?';
```
Comment (swaminator) [inline packages/amplify-console-hosting/constants/question-constants.js]: Do not ask the question for 'hosting serve' for manual deploys. Just open the app

File: packages/amplify-console-hosting/hosting/index.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+const questions = require('../modules/questions/question-generator');
+const ValidationError = require('../error/validation-error').default;
+const clientFactory = require('../utils/client-factory');
+const ora = require('ora');
+const tableUtis = require('../utils/table-utils');
+
+const HELP_INFO_PLACE_HOLDER =
+  'Manual deployment allows you to publish your web app to the Amplify Console without connecting a Git provider. Continuous deployment allows you to publish changes on every code commit by connecting your GitHub, Bitbucket, GitLab, or AWS CodeCommit repositories.';
+const REMOVE_ERROR_MESSAGE = 'There was an error removing the auth resource';
+const HOSTING_NOT_ENABLED = 'Amplify console hosting is not enabled!';
```
Comment (swaminator) [inline packages/amplify-console-hosting/hosting/index.js]: Amplify Console hosting is not enabled.

File: packages/amplify-console-hosting/hosting/index.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+const HOSTING_ALREADY_ENABLED =
+  'Amplify Console hosting has already been enabled';
+const META_FILE_VALIDATION =
+  'Amplify console hosting information is missed from amplify meta file';
```
Comment (swaminator) [inline packages/amplify-console-hosting/hosting/index.js]: When does this happen? This error isn't useful to the customer.

File: packages/amplify-console-hosting/hosting/index.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+const FRONTEND_EXISTED_WARNING =
+  'Branches has already been added to amplify app. Can not enable local host';
```
Comment (swaminator) [inline packages/amplify-console-hosting/hosting/index.js]: You have already connected branches to your Amplify Console app. Please visit the Amplify Console to manage your branches.

File: packages/amplify-console-hosting/utils/amplify-console-utils.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
+const { put } = require('request-promise');
+const fs = require('fs-extra');
+const ora = require('ora');
+
+const DEPLOY_ARTIFACTS_MESSAGE =
+  'Deploying build artifacts to the Amplify Console..';
+const DEPLOY_COMPLETE_MESSAGE = 'Deployment complete!';
+const DEPLOY_FAILURE_MESSAGE = 'Deploy failed!';
```
Comment (swaminator) [inline packages/amplify-console-hosting/utils/amplify-console-utils.js]: Deployment failed: <Reason>! Please report an issue on the Amplify Console GitHub issue tracker at https://github.com/aws-amplify/amplify-console/issues.

File: packages/amplify-console-hosting/utils/build-utils.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+const { spawn } = require('child_process');
+const chalk = require('chalk');
+
+const fs = require('fs-extra');
+const archiver = require('archiver');
+
+const DIR_NOT_FOUND_ERROR_MESSAGE = 'Please ensure your build path exist';
```
Comment (swaminator) [inline packages/amplify-console-hosting/utils/build-utils.js]: Please ensure your build artifacts path exists.

**Comment by garyleefight** [comment]
Fixed all

File: packages/amplify-console-hosting/hosting/templates/amplifyhosting-template.json [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        },
+        "appId": {
+            "Type": "String"
+        },
+        "type": {
+            "Type": "String"
+        }
+    },
+    "Conditions": {
+        "ShouldNotCreateEnvResources": {
```
[the remark this answers is not in the document]
Comment (garyleefight) [reply packages/amplify-console-hosting/hosting/templates/amplifyhosting-template.json]: Will remove

File: packages/amplify-console-hosting/amplify-plugin.json [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    "type": "category",
+    "commands": [
+        "add",
+        "configure",
+        "remove",
+        "publish",
+        "help",
+        "status",
+        "serve"
+    ],
```
[the remark this answers is not in the document]
Comment (garyleefight) [reply packages/amplify-console-hosting/amplify-plugin.json]: I remembered we allow user to use amplify hosting status and only show amplify hosting status as well as publish commands.
Will change the console part

**Pull request merged**